grpc-gcp/src/main/java/com/google/cloud/grpc/GcpMultiEndpointChannel.java

```diff
@@ -0,0 +1,402 @@
+   *
+   * @see #isShutdown()
+   * @since 1.0.0
+   */
+  @Override
+  public boolean isTerminated() {
+    return pools.values().stream().allMatch(GcpManagedChannel::isTerminated);
+  }
+
+  /**
+   * Initiates a forceful shutdown in which preexisting and new calls are cancelled. Although
+   * forceful, the shutdown process is still not instantaneous; {@link #isTerminated()} will likely
+   * return {@code false} immediately after this method returns.
+   *
+   * @return this
+   * @since 1.0.0
+   */
+  @Override
+  public ManagedChannel shutdownNow() {
+    pools.values().forEach(GcpManagedChannel::shutdownNow);
+    return this;
+  }
+
+  /**
+   * Waits for the channel to become terminated, giving up if the timeout is reached.
+   *
+   * @return whether the channel is terminated, as would be done by {@link #isTerminated()}.
+   * @since 1.0.0
+   */
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    long endTimeNanos = System.nanoTime() + unit.toNanos(timeout);
+    for (GcpManagedChannel gcpManagedChannel : pools.values()) {
+      if (gcpManagedChannel.isTerminated()) {
+        continue;
+      }
+      long awaitTimeNanos = endTimeNanos - System.nanoTime();
+      if (awaitTimeNanos <= 0) {
+        break;
+      }
+      gcpManagedChannel.awaitTermination(awaitTimeNanos, NANOSECONDS);
+    }
+    return isTerminated();
+  }
+
+  /**
+   * Check the value of {@link #ME_KEY} key in the {@link CallOptions} and if found use
+   * the MultiEndpoint with the same name for this call.
+   *
+   * <p>Create a {@link ClientCall} to the remote operation specified by the given {@link
+   * MethodDescriptor}. The returned {@link ClientCall} does not trigger any remote behavior until
+   * {@link ClientCall#start(Listener, Metadata)} is invoked.
+   *
+   * @param methodDescriptor describes the name and parameter types of the operation to call.
+   * @param callOptions runtime options to be applied to this call.
+   * @return a {@link ClientCall} bound to the specified method.
+   * @since 1.0.0
+   */
+  @Override
+  public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
+      MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
+    final String multiEndpointKey = callOptions.getOption(ME_KEY);
+    MultiEndpoint me = defaultMultiEndpoint;
+    if (multiEndpointKey != null) {
+      me = multiEndpoints.getOrDefault(multiEndpointKey, defaultMultiEndpoint);
+    }
+    return pools.get(me.getCurrentId()).newCall(methodDescriptor, callOptions);
+  }
+
+  /**
+   * The authority of the destination this channel connects to. Typically this is in the format
+   * {@code host:port}.
+   *
+   * This may return different values over time because MultiEndpoint may switch between endpoints.
+   *
+   * @since 1.0.0
+   */
+  @Override
+  public String authority() {
+    return pools.get(defaultMultiEndpoint.getCurrentId()).authority();
```

Comment:
Why always return the endpoint of the defaultMultiEndpoint? I assume `authority()` is a per call data instead of a per channel data, so shouldn't the current endpoint of the MultiEndpoint that actually gets used in a call be returned?

Reply:
The `defaultMultiEndpoint.getCurrentId()` will return an endpoint to be used for the call if the default MultiEndpoint is used. But unfortunately, for the `authority()` we don't have CallOptions to make sure which MultiEndpoint is desired.

Reply:
To me this looks like a bug, i.e., the `authority()` actually is not the authority used in a call. Can you fix this?

Reply:
The user of the multi-endpoint channel is aware that the endpoint for the call is chosen at the realtime. And the user provides all authorities for every endpoint while configuring the multi-endpoint. Do you know when this method may be used? I don't expect a user would want to use it and if they do they are aware of this behavior.

If a user really needs to fetch an authority for the current endpoint in a specific MultiEndpoint (not default) I can easily add a new method for that. WDYT?

Reply:
I do not know what is the use case of this method. But in general I think it is better to not assume users are aware of this method does not guarantee to return the endpoint actually chosen at runtime. So I suggest:
1. Add a new method that can correctly returns the endpoint;
2. If you want to keep this method, add a comment that explains the new method should be used instead of this hesitated one.

Reply:
Sounds good. Done.

I have to keep this method because it comes from io.grpc.Channel abstract class.